stl/src/filesystem.cpp

```diff
@@ -503,36 +499,124 @@ _Success_(return == __std_win_error::_Success) __std_win_error
     }
 
     FILE_DISPOSITION_INFO _Info{/* .Delete= */ TRUE};
-    if (SetFileInformationByHandle(_Handle._Get(), FileDispositionInfo, &_Info, sizeof(_Info))) {
+    if (SetFileInformationByHandle(reinterpret_cast<HANDLE>(_Handle), FileDispositionInfo, &_Info, sizeof(_Info))) {
+        return __std_win_error::_Success;
+    }
+
+    return __std_win_error{GetLastError()};
+}
+
+[[nodiscard]] __std_fs_remove_result __stdcall __std_fs_remove(_In_z_ const wchar_t* const _Target) noexcept {
+    // remove _Target without caring whether _Target is a file or directory
+    __std_win_error _Last_error;
+    bool _Able_to_change_attributes = false;
+
+    constexpr auto _Flags = __std_fs_file_flags::_Backup_semantics | __std_fs_file_flags::_Open_reparse_point;
+    _STD _Fs_file _Handle(_Target,
+        __std_access_rights::_Delete | __std_access_rights::_File_read_attributes
+            | __std_access_rights::_File_write_attributes,
+        _Flags, &_Last_error);
+    if (_Last_error == __std_win_error::_Success) {
+        _Able_to_change_attributes = true;
+    } else if (_Last_error == __std_win_error::_Access_denied) {
+        // change the underlying HANDLE of _Handle
+        // this might not be a good way of doing it (comment to be removed post review)
+        _Last_error = __std_fs_open_handle(&_Handle._Raw, _Target, __std_access_rights::_Delete, _Flags);
+        if (_Last_error != __std_win_error::_Success) {
+            return {false, _Last_error};
+        }
+    } else {
+        return {false, _Translate_not_found_to_success(_Last_error)};
+    }
+
+    // For Windows 10 1809 or later we have this flag -> FILE_DISPOSITION_FLAG_IGNORE_READONLY_ATTRIBUTE 0x10
+    // This flag also deletes read-only files
+    // NOTE: This is currently undocumented in MSDN. Refer WinBase.h for declarations
+
+    // The following bits are set here
+    // FILE_DISPOSITION_FLAG_DELETE, FILE_DISPOSITION_FLAG_POSIX_SEMANTICS,
+    // FILE_DISPOSITION_FLAG_IGNORE_READONLY_ATTRIBUTE
+    _File_disposition_info_ex _Info_ex{0x3 | 0x10};
+    constexpr auto _FileDispositionInfoExClass = static_cast<FILE_INFO_BY_HANDLE_CLASS>(21);
+    if (SetFileInformationByHandle(
+            reinterpret_cast<HANDLE>(_Handle._Get()), _FileDispositionInfoExClass, &_Info_ex, sizeof(_Info_ex))) {
         return {true, __std_win_error::_Success};
     }
 
-    return {false, __std_win_error{GetLastError()}};
+    _Last_error = __std_win_error{GetLastError()};
+    switch (_Last_error) {
+    // Windows versions older than 1809
+    case __std_win_error::_Invalid_parameter:
+    case __std_win_error::_Invalid_function:
+    case __std_win_error::_Not_supported:
+        break;
+    default:
+        // If we fail for reasons other than those mentioned above just return to the caller
+        return {false, _Last_error};
+    }
 
-#undef _SetFileInformationByHandle
+    // code path for versions older than 1809
+    _Last_error = _Set_delete_flag(_Handle._Raw);

```

Comment:
I believe this should be `_Handle._Get()` since it doesn't need to modify the handle in-place.

Reply:
okay

Reply:
Ah, thanks - I didn't look at the types closely! I think that this code is fine as-is then.